File: scripts_genomes/repetitiveness_1.m
function results = repetitiveness_1(user,genome)
% This script attempts to analyze genomic sequences for repetitive elements.
% Loads multiple FASTA-formatted sequence files defining a genome, then processes it using the repetitiveness algorithm.

% log file start, for in-process analysis.
fprintf(['** Genome : [[[' genome '[[[\n']);

workingDir             = ['../users/' user '/genomes/' genome '/'];
nmer_length            = 10;

[centromeres, chr_sizes, figure_details, annotations, ploidy_default] = Load_genome_information_1(workingDir,genome);

%% Determine number of chromosomes from figure_details.
num_chr    = 0;
chr_labels = [];
for i = 1:length(figure_details)
    if (figure_details(i).chr ~= 0)
        num_chr = num_chr+1;
        chr_names{figure_details(i).chr}  = figure_details(i).name;
    end;
end;

%% Determine reference genome FASTA file in use.
%  Read in and parse : "links_dir/main_script_dir/genome_specific/[genome]/reference.txt"
reference_file = [workingDir '/reference.txt'];
refernce_fid   = fopen(reference_file, 'r');
refFASTA       = fgetl(refernce_fid);
fclose(refernce_fid);
FastaName      = strrep(refFASTA,'.fasta','');

fprintf(['\n** number of chrs in figure = ' num2str(num_chr)]);

%% ====================================================================
% Parse sequences from FASTQ file, or load MAT files if already generated for this genome.
% ---------------------------------------------------------------------
if (exist([workingDir FastaName '.seq.mat'],'file') == 0) && (exist([workingDir FastaName '.seqRevCom.mat'],'file') == 0)
    fprintf('\n** Parsing sequences from FASTQ file.');
    SequenceData = fastaread([workingDir refFASTA ],'TrimHeaders', true);
    num_chr = length(SequenceData);

    %% Initialize the cell arrays needed to contain chromosome sequences.
    sequences         = cell(1,num_chr);
    rev_com_sequences = cell(1,num_chr);

    fprintf(['\n**\tLoading FASTA file: \"' refFASTA '\"\n']);
	for i = 1:length(SequenceData)
		sequences{i}         = SequenceData(i).Sequence;
		name                 = SequenceData(i).Header;
		fprintf([num2str(i) ' : ' name '\n']);
		name_strings         = textscan(name,'%s').';
		names{i}             = name_strings{1};
		rev_com_sequences{i} = rev_com(sequences{i});
    end;
    save([workingDir FastaName '.seq.mat']      ,'sequences', 'names');
    save([workingDir FastaName '.seqRevCom.mat'],'rev_com_sequences');
else
    fprintf('\n** Loading pre-determined sequences.');
    load([workingDir FastaName '.seq.mat']);
    load([workingDir FastaName '.seqRevCom.mat']);
end;

for i = 1:length(sequences)
    length_nts(i) = length(sequences{i});
end;

%% ====================================================================
% Tally incidence of nmers across genome.
% ---------------------------------------------------------------------
num_chr = length(sequences);
fprintf(['\n**\t' num2str(num_chr) ' chromosomes in genome sequence file.']);
if (exist([workingDir FastaName '.nmerCounts_' num2str(nmer_length) '.mat'],'file') == 0)
    fprintf('\n** Determining incidence of nmers across across genome.\n');
    nmer_counts = zeros(1,4^nmer_length,'int32');
    for chr = 1:num_chr
        Query_length = nmer_length;

	fprintf(['\n**\tAnalyzing ' num2str(nmer_length) 'mer repeats in : ' names{chr}{1}]);
%	disp(names{chr});

	fprintf(['\n**\t\tseq length = ' num2str(length_nts(chr))]);

	for Query_offset = 1:(length_nts(chr)-Query_length+1)
	    Query         = sequences{chr}(Query_offset:(Query_offset+Query_length-1));
	    rev_com_Query = rev_com_sequences{chr}((length_nts(chr)-Query_offset-Query_length+2):(length_nts(chr)-Query_offset+1));

	    % Add to counts of query sequence and reverse complement of query sequence.
	    [forward,err1] = find_nmer(Query);
	    [reverse,err2] = find_nmer(rev_com_Query);
	    if (err1 == false)
		nmer_counts(forward) = nmer_counts(forward)+1;
	    end;
	    if (err2 == false)
		nmer_counts(reverse) = nmer_counts(reverse)+1;
	    end;
	end;
    end;
    % full genome analysis for 5mers took:   27,253.760544 s = 454.229342 min = 7.570489 hr
    % and then I wrote 'find_nmers.m'...
    % full genome analysis for 10mers took:   1,093.824293 s =  18.230405 min = 0.303840 hr (~18 minutes)

    fprintf('\n** Saving incidence of nmers across genome.');
    save([workingDir FastaName '.nmerCounts_' num2str(nmer_length) '.mat'],'nmer_counts');
    clear counter forward reverse Query rev_com_Query Query_length;
    clear Query_offset err1 err2;
else
    fprintf('\n** Loading incidence of nmers across genome.');
    load([workingDir FastaName '.nmerCounts_' num2str(nmer_length) '.mat']);
end;

%% ====================================================================
% Determine distribution of repeats across genome.
% ---------------------------------------------------------------------
if (exist([workingDir FastaName '.nmerDists_' num2str(nmer_length) '.mat'],'file') == 0)
    fprintf('\n** Determining distribution of repeats across genome.');
    % Make a vectors to track incidence of repetitive sequences.
    clear repeats;
    repeats = cell(1,num_chr);
    for chr = 1:num_chr
        repeats{chr}(1:length_nts(chr)) = 0;
    end;
    
    % Examine sequence...
    for chr = 1:num_chr
        counter = 0;
        Query_length = nmer_length;
        fprintf(['\n**\tAnalyzing ' num2str(nmer_length) 'mer repeats in : ' names{chr}{1}]);
%		disp(names{chr});
        for Query_offset = 1:(length_nts(chr)-Query_length+1)
            counter = counter+1;
            Query = sequences{chr}(Query_offset:(Query_offset+Query_length-1));
            [forward, err1] = find_nmer(Query);
            if (err1 == false)
                if (~isempty(forward))
                    for j = Query_offset:(Query_offset+Query_length-1)
                        repeats{chr}(j) = repeats{chr}(j) + nmer_counts(forward);
                    end;
                end;
            end;
        end;
    end;
    fprintf('\n** Saving distribution of repeats across genome.');
    save([workingDir FastaName '.nmerDists_' num2str(nmer_length) '.mat'],'repeats');
    clear i j counter forward reverse Query Query_length Query_offset;
    clear rev_com_Query err1;
else
    fprintf('\n** Loading distribution of repeats across genome.');
    load([workingDir FastaName '.nmerDists_' num2str(nmer_length) '.mat']);
    for chr = 1:num_chr
        length_nts(chr) = length(repeats{chr});
    end;
    clear input_files sequences rev_com_sequences chr_names i;
end;


%% ====================================================================
% Output raw repetitiveness score per base pair.
% ---------------------------------------------------------------------
fprintf('\n** Outputing repetitiveness score file for genome.');

outFile = [workingDir FastaName '.repetitiveness.txt'];
fid = fopen(outFile,'w');
fprintf(fid,'## Repetitiveness score per bp location.\n');
fprintf(fid,'##\n');
fprintf(fid,'## columns = [chrName, bpCoordinate, repetitivenessScore]\n');
for chr = 1:num_chr
	fprintf(['\n**\t' names{chr}{1}]);
%	disp(names{chr}{1});
	for bp = 1:length_nts(chr)
		repetitiveness_value = repeats{chr}(bp);
		fprintf(fid, names{chr}{1} );
		fprintf(fid,['\t' num2str(bp) '\t' num2str(repetitiveness_value) '\n']);
 	end;
end;
fclose(fid);
fprintf('\n**\n');


%% ====================================================================
% Smooth repetitiveness score & output
% ---------------------------------------------------------------------
%for smooth_level = 1:20
%    sigma = smooth_level*1000;
%    fprintf(['\nSmoothing raw repetitiveness plot (sigma = ' num2str(sigma) ') [']);
%    for chr = 1:num_chr
%	fprintf(num2str(chr));
%	% smooth_gaussian(data,sigma,size)
%	%    data  : input vector with raw data.
%	%    sigma : standard deviation of the gaussian distribution used in the smoothing.
%	%    size  : size of vector over which smoothing function is applied.   (2-3 sigmas is usually good.)
%	s1{chr} = smooth_gaussian(repeats{chr},sigma,sigma*2);
%    end;
%    fprintf(']');
%
%    outFile = [workingDir FastaName '.repetitiveness.sigma_' num2str(sigma) '.txt'];
%    fid = fopen(outFile,'w');
%    fprintf(fid,'## Repetitiveness score per bp location.\n');
%    fprintf(fid,['## Raw scores are Gaussian smoothed with sigma = ' num2str(sigma) '.\n']);
%    fprintf(fid,'##\n');
%    fprintf(fid,'## columns = [chrName, bpCoordinate, repetitivenessScore]\n');
%    for chr = 1:num_chr
%	fprintf(['\n**\t' names{chr}{1}]);
%	for bp = 1:length_nts(chr)
%	    repetitiveness_value = repeats{chr}(bp);
%	    fprintf(fid,[names{chr} '\t' num2str(bp) '\t' num2str(repetitiveness_value) '\n']);
%	end;
%    end;
%    fclose(fid);
%    fprintf('\n**\n');
%end;

end

File: scripts_genomes/repetitiveness_1.py
# Input arguments:
#	1) user     : Name of user account installing the genome file.
#	2) genome   : Name of the genome.
#	3) main_dir : Root directory.
#	4) logfile  : Path and file name of output log file.
#
import string, sys, re, time;
userName    = sys.argv[1];
genomeName  = sys.argv[2];
main_dir    = sys.argv[3];
logName     = sys.argv[4];

nmer_length = 10;

with open(logName, "a") as myfile:
	myfile.write("\t\t\t*================================================================*\n")
	myfile.write("\t\t\t| Log of 'repetitiveness_1.py'                                   |\n")
	myfile.write("\t\t\t*----------------------------------------------------------------*\n")


#------------------------------------------------------------------------------------------------------------
# Calculates position in nmer count vector and generates an error condition if the test sequence contains non-ATCG characters.
def find_nmer(seq):
	steps = len(seq);
	nt    = [0]*steps;
	err   = 'false';
	seq   = seq.upper();
	for index in range(steps):
		if seq[index] == 'A':
			nt[index] = 0;
		elif seq[index] == 'T':
			nt[index] = 1;
		elif seq[index] == 'C':
			nt[index] = 2;
		elif seq[index] == 'G':
			nt[index] = 3;
		else:
			err = 'true';
	nmer = 0;  # vectors in python are 0-rooted.
	for index in range(steps):
		nmer += nt[index]*4**index;
	return [nmer,err];

#------------------------------------------------------------------------------------------------------------
# Recursively reverse the input string.
def reverse(text):
	if len(text) <= 1:
		return text;
	return reverse(text[1:]) + text[0];

#------------------------------------------------------------------------------------------------------------
# Generate the reverse complement sequence of an input DNA sequence string.
def rev_com(seq):
	seq         = seq.upper();
	rev_seq     = reverse(seq);
	rev_com_seq = '';
	for index in range(len(seq)):
		if   rev_seq[index] == 'A':
			rev_com_seq += 'T';
		elif rev_seq[index] == 'T':
			rev_com_seq += 'A';
		elif rev_seq[index] == 'C':
			rev_com_seq += 'G';
		elif rev_seq[index] == 'G':
			rev_com_seq += 'C';
		else:
			rev_com_seq += 'n';
	return rev_com_seq;

#------------------------------------------------------------------------------------------------------------


# Initialize time counter and log file section.
t0 = time.clock();


#============================================================================================================
# Determine name of reformatted reference FASTA file.
#------------------------------------------------------------------------------------------------------------
# Find name of genome FASTA file for species being examined.
#     Read in and parse : "Ymap_root/users/[userName]/genomes/[genomeName]/[genome]/reference.txt"
workingDir     = main_dir + 'users/' + userName + '/genomes/' + genomeName + '/';
reference_file = workingDir + 'reference.txt';
refFile        = open(reference_file,'r');
refFASTA       = refFile.read().strip();
refFile.close();

# Generate the name of the reformatted FASTA file: "test.fasta" => "test.2.fasta"
refFASTA       = refFASTA.replace(".fasta",".2.fasta");
with open(logName, "a") as myfile:
	myfile.write("\n\t\t\tFASTA reformatted to single-line entries : " + refFASTA);


#============================================================================================================
# Determine which chromosomes from reference FASTA are being examined.
#------------------------------------------------------------------------------------------------------------
# Look up chromosome name strings for genome in use.
#     Read in and parse : "Ymap_root/users/[userName]/genomes/[genomeName]/figure_definitions.txt"
figureDefinition_file  = workingDir + '/figure_definitions.txt';
figureDefinitionFile   = open(figureDefinition_file,'r');
figureDefinitionData   = figureDefinitionFile.readlines();
#	Example lines in figureDefinition_file:
#		Chr  Use   Label   Name                         posX   posY   width   height
#		1    1     Chr1    Ca21chr1_C_albicans_SC5314   0.15   0.8    0.8     0.0625
#		2    1     Chr2    Ca21chr2_C_albicans_SC5314   0.15   0.7    *       0.0625
#		0    0     Mito    Ca19-mtDNA                   0.0    0.0    0.0     0.0

with open(logName, "a") as myfile:
	myfile.write("\n\t\t\t\tDetermining number of chromosomes of interest in genome.");

# Determine the number of chromosomes of interest in genome.
chrName_maxcount = 0;
for line in figureDefinitionData:
	line_parts = string.split(string.strip(line));
	chr_num = line_parts[0];
	if chr_num.isdigit():
		chr_num    = int(float(line_parts[0]));
		chr_use    = int(float(line_parts[1]));
		chr_label  = line_parts[2];
		chr_name   = line_parts[3];
		if chr_num > 0:
			if chr_num > chrName_maxcount:
				chrName_maxcount = chr_num;
figureDefinitionFile.close();

# Pre-allocate chrName_array
chrName = [];
for x in range(0, chrName_maxcount):
	chrName.append([]);

with open(logName, "a") as myfile:
	myfile.write("\n\t\t\t\tGathering name strings for chromosomes.");

# Gather name strings for chromosomes, in order.
figureDefinitionFile  = open(figureDefinition_file,'r');
for line in figureDefinitionData:
	line_parts = string.split(string.strip(line));
	chr_num = line_parts[0];
	if chr_num.isdigit():
		chr_num    = int(float(line_parts[0]));
		chr_use    = int(float(line_parts[1]));
		chr_label  = line_parts[2];
		chr_name   = line_parts[3];
		if chr_num != 0:
			chrName[int(float(chr_num))-1] = chr_name;
			with open(logName, "a") as myfile:
				myfile.write("\n\t\t\t\t\tChr" + str(chr_num) + " = " + chr_name);
figureDefinitionFile.close();

# Put the chromosome count into a smaller name for later use.
chrCount = chrName_maxcount;


#============================================================================================================
# Process used chromosomes into k-mer counts array.
#------------------------------------------------------------------------------------------------------------
# Initialize nmer_counts vector, zeros of length 4^nmer_length.
nmer_counts = [0]*(4**nmer_length);

# Open reformatted FASTA file.
FASTA_file = workingDir + refFASTA;
FASTA_data = open(FASTA_file,'r');

# Reset timer.
t0a = time.clock();
t1  = t0a;
with open(logName, "a") as myfile:
	myfile.write("\n\t\t\t#### Start tallying nmers.\n");

# Process reformatted FASTA file, entry by entry, to collect nmer counts.
while True:
	# FASTA entries are pairs of lines with the following structure.
	#    >Ca_a.chr1 (9638..10115) (478bp) [*]
	#    ATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGC
	line1 = FASTA_data.readline();
	line2 = FASTA_data.readline();
	if not line2:
		break  # EOF
	first_char = line1[:1];
	old_chr_name = '';
	if first_char == ">":
		# First line is header to FASTA entry, so file contents are formatted properly.
		# Determine chromosome/contig name by isolating the first space-delimited string, then removing the header character ">".
		line_parts = string.split(string.strip(line1));
		chr_name   = line_parts[0];
		chr_name   = chr_name.replace(">","");

		# If the current chromosome is different than the last one, output log entry.
		if (chr_name <> old_chr_name):
			with open(logName, "a") as myfile:
				myfile.write("\t\t\t####\t" + str(time.clock() - t1) + " seconds for this chr.\n");
				t1 = time.clock();
				if chr_name in chrName:
					myfile.write("\t\t\t#### Tallying nmers of: " + chr_name + "\n");
				else:
					myfile.write("\t\t\t#### Skipping '" + chr_name + "' because it was deselected.\n");

		# If the current chromosome is one of those in use...
		if chr_name in chrName:
			# Run along chromosome coordinates, from start(0) to end (len(line2)-nmer_length+1), such that a nmer can be examined starting at each coordinate.
			# For each coordinate, add to k-mer count vector "nmer_counts".
			count = 0;
			for index in range(0, len(line2)-nmer_length+1):
				count += 1;
				# current nmer string.
				test_string      = line2[index:(index+nmer_length)];
				# determine nmer_counts vector position for the test_string and its reverse complement.
				results_1        = find_nmer(test_string);
				results_2        = find_nmer(rev_com(test_string));
				forward_nmer_num = results_1[0];
				forward_nmer_err = results_1[1];
				reverse_nmer_num = results_2[0];
				if forward_nmer_err == 'false':
					# If test_string is a valid DNA sequence, increment the nmer_counts vector position for the string and its reverse complement.
					nmer_counts[forward_nmer_num] += 1;
					nmer_counts[reverse_nmer_num] += 1;
		old_chr_name = chr_name;
FASTA_data.close();

# Reset timer.
with open(logName, "a") as myfile:
	if old_chr_name in chrName:
		myfile.write("\t\t\t####\t" + str(time.clock() - t1) + " seconds for this chr.\n");
	myfile.write("\t\t\t####\n");
	myfile.write("\t\t\t#### " + str(time.clock() - t0a) + " seconds to talley incidence of nmers across genome.\n");
	myfile.write("\t\t\t####\n");
t0b = time.clock();
t1  = t0b;
with open(logName, "a") as myfile:
	myfile.write("\t\t\t#### Start generating repetitiveness score across genome.\n");


#============================================================================================================
# Output header lines for repetititveness file output.
#------------------------------------------------------------------------------------------------------------
print '## Repetitiveness score per bp location.';
print '##';
print '## columns = [chrName, bpCoordinate, repetitivenessScore]';


#============================================================================================================
# Process k-mer array into repetitiveness scores for each coordinate, then generate output lines.
#------------------------------------------------------------------------------------------------------------
FASTA_data = open(FASTA_file,'r');
# chrName[] contains names of used chromosomes.
while True:
	# FASTA entries are pairs of lines with the following structure.
	#    >Ca_a.chr1 (9638..10115) (478bp) [*]
	#    ATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGC
	line1 = FASTA_data.readline();
	line2 = FASTA_data.readline();
	if not line2:
		break  # EOF
	first_char = line1[:1];
	old_chr_name = '';
	if first_char == ">":
		# First line is header to FASTA entry, so file contents are formatted properly.
		# Determine chromosome/contig name by isolating the first space-delimited string, then removing the header character ">".
		line_parts = string.split(string.strip(line1));
		chr_name   = line_parts[0];
		chr_name   = chr_name.replace(">","");

		# If the current chromosome is different than the last one, output log entry.
		if (chr_name <> old_chr_name):
			with open(logName, "a") as myfile:
				myfile.write("\t\t\t####\t" + str(time.clock() - t1) + " seconds for this chr.\n");
			t1 = time.clock();
			with open(logName, "a") as myfile:
				myfile.write("\t\t\t#### Outputing repetitiveness scores for: " + chr_name + "\n");

		# If the current chromosome is one of those in use...
		if chr_name in chrName:
			# Run along chromosome coordinates, from start (0) to end (len(line2)-1), such that all nmers overlapping the coordinate can be examined.
			for index in range(0, len(line2)-1):
				score_sum = 0;
				for index_offset in range(-nmer_length+1,1):
					# current nmer string.
					if (index+index_offset >= 0) and (index+index_offset+nmer_length-1 <= len(line2)-1):
						test_string      = line2[(index+index_offset):(index+index_offset+nmer_length)];
						# determine nmer_counts vector position for the test_string. The reverse-complement was dealt with in the tallying step previous.
						results_1        = find_nmer(test_string);
						forward_nmer_num = results_1[0];
						forward_nmer_err = results_1[1];
						if forward_nmer_err == 'false':
							# If test_string is a valid DNA sequence, add to score_sum.
							score_sum += nmer_counts[forward_nmer_num];
				# Output repetitiveness score line to file: [chrName, bpCoordinate, repetitivenessScore]
				print chr_name + '\t' + str(index+1) + '\t' + str(score_sum);
		old_chr_name = chr_name;


#============================================================================================================
# Conclude log outputs.
#------------------------------------------------------------------------------------------------------------
with open(logName, "a") as myfile:
	myfile.write("\t\t\t####\t" + str(time.clock() - t1) + " seconds for this chr.\n");
	myfile.write("\t\t\t####\n");
	myfile.write("\t\t\t#### Total time for computation of genome repetitiveness = " + str(time.clock() - t0) + "\n");
	myfile.write("\t\t\t####\n");
